File: example/src/main/kotlin/com/xebia/functional/langchain4k/auto/Employee.kt
package com.xebia.functional.langchain4k.auto

import com.xebia.functional.auto.ai
import com.xebia.functional.auto.getOrElse
import kotlinx.serialization.Serializable

@Serializable
data class Employee(val firstName: String, val lastName: String, val age: Int, val position: String, val company: Company)

@Serializable
data class Address(val street: String, val city: String, val country: String)

@Serializable
data class Company(val name: String, val address: Address)

suspend fun main() {
    ai {
      val complexPrompt =
        """|
                 |Provide made up information for an Employee that includes their first name, last name, age, position, and their company's name and address (street, city, and country).
                 |Use the information provided.
              """.trimMargin()

      val employeeData: Employee = ai(complexPrompt)

      println(
        "Employee Information:\n\n" +
          "Name: ${employeeData.firstName} ${employeeData.lastName}\n" +
          "Age: ${employeeData.age}\n" +
          "Position: ${employeeData.position}\n" +
          "Company: ${employeeData.company.name}\n" +
          "Address: ${employeeData.company.address.street}, ${employeeData.company.address.city}, ${employeeData.company.address.country}"
      )
    }.getOrElse { println(it) }
}

File: examples/scala/src/main/scala/com/xebia/functional/xef/scala/auto/Employee.scala
package com.xebia.functional.xef.scala.auto

import com.xebia.functional.xef.scala.auto.*
import io.circe.Decoder

private final case class Employee(firstName: String, lastName: String, age: Int, position: String, company: Company)
    derives SerialDescriptor,
      Decoder

private final case class Address(street: String, city: String, country: String) derives SerialDescriptor, Decoder

private final case class Company(name: String, address: Address) derives SerialDescriptor, Decoder

@main def runEmployee: Unit =
  ai {
    val complexPrompt =
      """
        |Provide made up information for an Employee that includes their first name, last name, age, position,
        |and their company's name and address (street, city, and country).
        |Use the information provided.
      """.stripMargin
    val employeeData = prompt[Employee](complexPrompt)
    println(
      s"""
         |Employee Information:
         |Name: ${employeeData.firstName} ${employeeData.lastName}
         |Age: ${employeeData.age}
         |Position: ${employeeData.position}
         |Company: ${employeeData.company.name}
         |Address: ${employeeData.company.address.street},${employeeData.company.address.city}, ${employeeData.company.address.country}
      """.stripMargin
    )
  }.getOrElse(ex => println(ex.getMessage))
